kafka-rest/src/main/java/io/confluent/kafkarest/controllers/SchemaManagerImpl.java

```diff
@@ -0,0 +1,348 @@
+        SchemaCacheLoader keyCacheLoader, SchemaCacheLoader valueCacheLoader) {
+      this.keyCacheLoader = requireNonNull(keyCacheLoader);
+      this.valueCacheLoader = requireNonNull(valueCacheLoader);
+    }
+
+    @Override
+    public RegisteredSchema load(GetSchemaParams params) throws Exception {
+      if (params.isKey()) {
+        return keyCacheLoader.load(params);
+      } else {
+        return valueCacheLoader.load(params);
+      }
+    }
+  }
+
+  private static final class SchemaCacheLoader
+      extends CacheLoader<GetSchemaParams, RegisteredSchema> {
+    private final String schemaType;
+    private final AbstractKafkaSchemaSerDe serializer;
+    private final SchemaProvider schemaProvider;
+    private final Function<String, String> subjectFunction;
+
+    private SchemaCacheLoader(
+        String schemaType,
+        AbstractKafkaSchemaSerDe serializer,
+        SchemaProvider schemaProvider,
+        Function<String, String> subjectFunction) {
+      this.schemaType = requireNonNull(schemaType);
+      this.serializer = requireNonNull(serializer);
+      this.schemaProvider = requireNonNull(schemaProvider);
+      this.subjectFunction = requireNonNull(subjectFunction);
+    }
+
+    @Override
+    public RegisteredSchema load(GetSchemaParams params) throws Exception {
+      switch (params.getParams().getOperation()) {
+        case SCHEMA_ID:
+          return getSchemaById(params.getParams().getSchemaId().getSchemaId());
+
+        case RAW_SCHEMA:
+          return parseSchema(
+              params.getParams().getRawSchema().getTopicName(),
+              params.getParams().getRawSchema().getRawSchema());
+
+        default:
+          throw new AssertionError(
+              String.format("Unexpected enum constant: %s", params.getParams().getOperation()));
+      }
+    }
+
+    private RegisteredSchema getSchemaById(int schemaId) {
+      ParsedSchema schema;
+      try {
+        schema = serializer.getSchemaById(schemaId);
+      } catch (IOException | RestClientException e) {
+        throw new SerializationException("Error when fetching schema by id.", e);
+      }
+      if (!schemaType.equals(schema.schemaType())) {
+        throw new SerializationException(
+            String.format(
+                "Fetched schema has the wrong type. Expected `%s' but got `%s'.",
+                schemaType,
+                schema.schemaType()));
+      }
+      return RegisteredSchema.create(schemaId, schema);
+    }
+
+    private RegisteredSchema parseSchema(String topicName, String rawSchema)
+        throws RestClientException, IOException {
+      ParsedSchema schema = schemaProvider.parseSchema(rawSchema, emptyList())
+          .orElseThrow(() -> new SerializationException("Error when parsing raw schema."));
+      int schemaId = serializer.register(subjectFunction.apply(topicName), schema);
+      return RegisteredSchema.create(schemaId, schema);
+    }
+  }
+
+  @AutoValue
```

Comment:
fancy.

is `GetSchemaParams` a good name? maybe just `SchemaParams`?

Reply:
Obsolete.